Document Deployment Runtime Config (#570)
crossplane/docs: content/master/concepts/providers.md

### Controller configuration

{{< hint "important" >}}
<!-- vale write-good.Passive = NO -->
<!-- vale gitlab.FutureTense = NO -->
The `ControllerConfig` type was deprecated in v1.11 and will be removed in
a future release.
<!-- vale write-good.Passive = YES -->
<!-- vale gitlab.FutureTense = YES -->

[`DeploymentRuntimeConfig`]({{<ref "#runtime-configuration" >}}) is the

> replacement for Controller configuration and is available in v1.14+.

Reviewer: ```suggestion
replacement for Controller configuration and is available in v1.14.
```

I think the + is implied.

Each Provider determines their supported set of `args`.

### Runtime configuration

{{<hint "important" >}}
`DeploymentRuntimeConfigs` is a beta feature. 

> It's on by default, and you could disable it by passing

Reviewer: ```suggestion
It's on by default, and you can disable it by passing
```

Nit: Not sure whether this is more correct, but to me it reads more naturally.

`--enable-deployment-runtime-configs=false` to the Crossplane deployment.
{{< /hint >}}

Runtime configuration is a generalized mechanism for configuring runtime for
Crossplane packages with a runtime, namely `Providers` and `Functions`. It

> replaces the deprecated `ControllerConfig` type and is available in v1.14+.

Reviewer: Perhaps "configuring the runtime". You could also just say "for configuring Crossplane packages with a runtime".

With its default configuration, Crossplane uses Kubernetes Deployments to
deploy runtime for packages, more specifically, a controller for a `Provider`
or a gRPC server for a `Function`. It's possible to configure the runtime
manifest by applying a `DeploymentRuntimeConfig` and referencing it in the
`Provider` or `Function` object.

Different from `ControllerConfig`, `DeploymentRuntimeConfig` embed the whole
Kubernetes Deployment spec, which allows for more flexibility in configuring
the runtime. Refer to the [design document](https://github.com/crossplane/crossplane/blob/2c5e7f07ba9e3d83d1c85169bbde685de8514ab8/design/one-pager-package-runtime-config.md)

> for more details.

Reviewer: This seems like it could make a good `{{note}}`.

As an example, to enable the external secret stores alpha feature for a `Provider`
by adding the `--enable-external-secret-stores` argument to the controller,
one can apply the following:

```yaml
apiVersion: pkg.crossplane.io/v1
kind: Provider
metadata:
  name: provider-gcp-iam
spec:
  package: xpkg.upbound.io/upbound/provider-gcp-iam:v0.37.0
  runtimeConfigRef:
    name: enable-ess
---
apiVersion: pkg.crossplane.io/v1beta1
kind: DeploymentRuntimeConfig
metadata:
  name: enable-ess
spec:
  deploymentTemplate:
    spec:
      selector: {}
      template:
        spec:
          containers:
            - name: package-runtime
              args:
                - --enable-external-secret-stores
```

> Please note that, the packages manager uses  `package-runtime` as the name of

Reviewer: ```suggestion
Please note that the packages manager uses  `package-runtime` as the name of
```

I don't think a comma is needed here.

the runtime container. When you use a different container name, the package
manager introduces it as a sidecar container instead of modifying the
package runtime container.

<!-- vale write-good.Passive = NO -->
The package manager is opinionated about some fields to ensure
<!-- vale write-good.Passive = YES -->
the runtime is working and overlay them on top of the values

> in the runtime configuration. For example, it defaults the replica count
> to 1 if not set or override the label selectors to make sure the Deployment
> and Service match. It also injects any necessary environment variables,
> ports as well as volumes and volume mounts.

Reviewer: ```suggestion
in the runtime configuration. For example, it defaults the replica count
to 1 if not set and overrides the label selectors to make sure the Deployment
and Service match. It also injects any necessary environment variables,
ports as well as volumes and volume mounts.
```

Grammar nit. 😄

The `Provider` or `Functions`'s `spec.runtimeConfigRef.name` field defaults
<!-- vale write-good.Passive = NO -->
to value `default`, which means the default runtime configuration is used if

> <!-- vale write-good.Passive = YES -->

Reviewer: You could avoid angering Vale here by writing "Crossplane uses the default runtime configuration" to make it active voice.

not specified. Crossplane ensures there is always a default runtime

> configuration in the cluster, but not change it if it already exists. This

Reviewer: ```suggestion
configuration in the cluster, but won't change it if it already exists. This
```

allows users to customize the default runtime configuration to their needs.

{{<hint "tip" >}}
<!-- vale gitlab.SubstitutionWarning = NO -->
Since `DeploymentRuntimeConfig` uses the same schema as Kubernetes `Deployment`
<!-- vale gitlab.SubstitutionWarning = YES -->
spec, you may need to pass empty values to bypass the schema validation.
For example, if you just want to change the `replicas` field, you would

> need to pass the following and it's fine:

Reviewer: ```suggestion
need to pass the following:
```

I think the fact this is fine/valid is implied.

```yaml
apiVersion: pkg.crossplane.io/v1beta1
kind: DeploymentRuntimeConfig
metadata:
  name: multi-replicas
spec:
  deploymentTemplate:
    spec:
      replicas: 2
      selector: {}
      template: {}
```

{{< /hint >}}

#### Configuring runtime deployment spec

> Using the Deployment spec provided in the `DeploymentRuntimeConfig` as base,

Reviewer: ```suggestion
Using the Deployment spec provided in the `DeploymentRuntimeConfig` as a base,
```
